### PR TITLE
Bump PL to 0.46.0-dev88

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
Bump the dep version of PL to 0.46.0-dev88 to benefit from the latest changes in Op2 and Abstract types.

[sc-129534]

### CODE REVIEW

**Review by paul0403 — APPROVED** [review]

**Comment by paul0403** [comment]
#3166

File: frontend/catalyst/decomposition/type_utils.py [hunk 1]
```diff
@@ -135,36 +135,49 @@ def get_dummy_values_for_arg(arg):
...
-            return qp.wires.Wires(range(len(node)))
+        if isinstance(node, (qp.typing.AbstractWires, qp.wires.Wires)):
+            return _placeholder_wires(len(node))
         else:
             return node
 
 
-def _replace_op_abstract_wires_with_concrete_wires(op2):
+def _placeholder_wires(num_wires):
+    """Return `num_wires` placeholder wires, labelled with negative integers."""
```
Comment (maliasadi) [inline frontend/catalyst/decomposition/type_utils.py]:
```suggestion
    """Return ``num_wires`` placeholder wires, labelled with negative integers."""
```

File: frontend/catalyst/decomposition/type_utils.py [hunk 2]
```diff
@@ -135,36 +135,49 @@ def get_dummy_values_for_arg(arg):
...
+    return qp.wires.Wires(range(-1, -num_wires - 1, -1))
+
+
+def _replace_op_wires_with_placeholder_wires(op2):
     """
     Given an Operator2 instance, return a copy of the same instance but with all fields whose value
-    is an `AbstractWires` replaced with concrete `Wires`.
+    is an `AbstractWires` or a concrete `Wires` replaced with placeholder `Wires`.
```
Comment (maliasadi) [inline frontend/catalyst/decomposition/type_utils.py]:
```suggestion
    is an ``AbstractWires`` or a concrete ``Wires`` replaced with placeholder ``Wires``.
```

**Comment by andrijapau** [comment]
Happy to close [mine](https://github.com/PennyLaneAI/catalyst/pull/3166) as you have more movement on yours! 😄 

Although, I thought changes to `uid.py` were needed as well as a regression test? 🤔

**Comment by maliasadi** [comment]
@andrijapau haha could you review the changes here? would you like to cherry-pick your work to this branch so we can also get LLL green ASAP with new checks?

**Comment by andrijapau** [comment]
> @andrijapau haha could you review the changes here? would you like to cherry-pick your work to this branch so we can also get LLL green ASAP?

Sure. Can we use the urgent label to speed things up?

File: .dep-versions [hunk 3]
```diff
@@ -8,7 +8,7 @@ enzyme=v0.0.238
 
 # For a custom PL version, update the package version here and at
 # 'doc/requirements.txt'
-pennylane=0.46.0.dev86
+pennylane=0.46.0-dev88
```
Comment (maliasadi) [inline .dep-versions]:
```suggestion
pennylane=0.46.0.dev88
```

File: doc/requirements.txt [hunk 4]
```diff
@@ -34,4 +34,4 @@ lxml_html_clean
 --extra-index-url https://test.pypi.org/simple/
 pennylane-lightning-kokkos==0.46.0-dev27
 pennylane-lightning==0.46.0-dev27
-pennylane==0.46.0.dev86
+pennylane==0.46.0-dev88
```
Comment (maliasadi) [inline doc/requirements.txt]:
```suggestion
pennylane==0.46.0.dev88
```

File: frontend/catalyst/decomposition/type_utils.py [hunk 5]
```diff
@@ -135,36 +135,49 @@ def get_dummy_values_for_arg(arg):
...
-    is an `AbstractWires` replaced with concrete `Wires`.
+    is an ``AbstractWires`` or a concrete ``Wires`` replaced with placeholder ``Wires``.
     """
     new_op = copy.deepcopy(op2)
     for wire_arg in new_op.wire_argnames:
-        if isinstance(new_op.arguments[wire_arg], qp.typing.AbstractWires):
-            num_wires = len(new_op.arguments[wire_arg])
-            new_op.arguments[wire_arg] = qp.wires.Wires(range(-1, -num_wires - 1, -1))
+        wire_val = new_op.arguments[wire_arg]
+        if isinstance(wire_val, (qp.typing.AbstractWires, qp.wires.Wires)):
```
Comment (andrijapau) [inline frontend/catalyst/decomposition/type_utils.py]: Any reason we still need this guard? Can't we guarantee that we'll either get Wires or AbstractWires here *always*?

File: frontend/catalyst/decomposition/type_utils.py [hunk 6]
```diff
@@ -135,36 +135,49 @@ def get_dummy_values_for_arg(arg):
...
-        if isinstance(node, qp.typing.AbstractWires):
-            return qp.wires.Wires(range(len(node)))
+        if isinstance(node, (qp.typing.AbstractWires, qp.wires.Wires)):
+            return _placeholder_wires(len(node))
         else:
             return node
 
 
-def _replace_op_abstract_wires_with_concrete_wires(op2):
+def _placeholder_wires(num_wires):
```
Comment (andrijapau) [inline frontend/catalyst/decomposition/type_utils.py]:
```suggestion
def _placeholder_wires(num_wires: int) -> qp.wires.Wires:
```

File: frontend/catalyst/decomposition/type_utils.py [hunk 7]
```diff
@@ -135,36 +135,49 @@ def get_dummy_values_for_arg(arg):
     raise TypeError(f"Unexpected type in container when creating dummy values: {type(arg)}")
 
 
-def replace_abstract_wires_with_concrete_wires(node):
+def replace_wires_with_placeholder_wires(node):
```
Comment (andrijapau) [inline frontend/catalyst/decomposition/type_utils.py:144]: If we have JAX available, can't we just use `tree_map` or something like that ([source](https://docs.jax.dev/en/latest/_autosummary/jax.tree.map.html))?

File: frontend/catalyst/decomposition/type_utils.py (hunk 5 above)
Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py]: good catch! the guard is actually redundant

Fixed [8e3cbfe](https://github.com/PennyLaneAI/catalyst/pull/3164/commits/8e3cbfe26455f8ffca70ea8f730011c247b52775)

File: frontend/catalyst/decomposition/type_utils.py (hunk 6 above)
Comment (andrijapau) [inline frontend/catalyst/decomposition/type_utils.py]: Conveys the purpose a little better IMO
```suggestion
def _create_placeholder_wires(num_wires):
```

Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py]: isn't `create_` a bit redundant?

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py]: Yeah, I can see your point. Happy to resolve.

File: frontend/catalyst/decomposition/type_utils.py (hunk 7 above)
Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py:144]: Nice suggestion! I'm working on it now

Fixed [8e3cbfe](https://github.com/PennyLaneAI/catalyst/pull/3164/commits/8e3cbfe26455f8ffca70ea8f730011c247b52775)

File: frontend/catalyst/decomposition/type_utils.py (hunk 6 above)
Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py]: not needed anymore [8e3cbfe](https://github.com/PennyLaneAI/catalyst/pull/3164/commits/8e3cbfe26455f8ffca70ea8f730011c247b52775)

File: frontend/catalyst/decomposition/type_utils.py (hunk 7 above)
Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:144]: Nice!

File: frontend/catalyst/decomposition/type_utils.py [hunk 8]
```diff
@@ -135,40 +135,28 @@ def get_dummy_values_for_arg(arg):
...
+    return isinstance(node, (qp.typing.AbstractWires, qp.wires.Wires))
 
 
-def _replace_op_abstract_wires_with_concrete_wires(op2):
-    """
-    Given an Operator2 instance, return a copy of the same instance but with all fields whose value
-    is an `AbstractWires` replaced with concrete `Wires`.
+def replace_wires_with_placeholder_wires(node):
     """
-    new_op = copy.deepcopy(op2)
```
Comment (andrijapau) [inline frontend/catalyst/decomposition/type_utils.py:160]: Was the deepcopy crucial here?

Comment (paul0403) [reply frontend/catalyst/decomposition/type_utils.py:160]: To not actually change the original op2 instance. This is just for creating dummies to the qjit instance that compiles the rules and should not touch the main user circuit's qjit.

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:160]: Okay, then we should preserve that and add a regression test if its loading bearing? @maliasadi

Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py:160]: I agree with @andrijapau that we need to add a test to check this :+1: 

@andrijapau is this the test you've added in your branch? would you like to contribute it to this PR?

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:160]: > @andrijapau is this the test you've added in your branch? 

No it's not ... 😅

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:160]: I can write a regression test though, I'll add it to `test_decomposition.py` under `TestGenericUtilities`.

Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py:160]: Go for it! :shipit:

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:160]: Let me know if this is insufficient: https://github.com/PennyLaneAI/catalyst/pull/3164/commits/212bd6df5e3bf105afc3d9daa348810f07a1a832

Comment (andrijapau) [reply frontend/catalyst/decomposition/type_utils.py:160]: `Operator2._unflatten` will create a new `Operator2` instance btw,
```python
    @classmethod
    def _unflatten(cls, data: Iterable[Any], metadata: Hashable):
        """Recreate an operation from its serialized format.

        Args:
            data: the dynamic component of the operation
            metadata: the static component of the operation.

        The output of ``Operator2._flatten`` and the class type must be sufficient to reconstruct the original
        operation with ``Operator2._unflatten``.

        **Example:**

        >>> op = qp.PauliRot(1.5, "XY", wires=[0, 1])
        >>> op._flatten()
        (([1.5], [Wires([0, 1])], []), ('XY',))
        >>> qp.PauliRot._unflatten(*op._flatten())
        PauliRot(theta=1.5, pauli_word=XY, wires=[0, 1])
        """
        ...

        with QueuingManager.stop_recording(), pause():
            return cls(**args)

```

Comment (maliasadi) [reply frontend/catalyst/decomposition/type_utils.py:160]: Thanks @andrijapau! Added the logic to utils [011585f](https://github.com/PennyLaneAI/catalyst/pull/3164/commits/011585f154575c77ae32701faf4249d7f3787394)

**Review by andrijapau — APPROVED** [review]
